File: src/main/java/org/tobi29/scapes/engine/gui/GuiComponentEvent.java
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.gui;

import org.tobi29.scapes.engine.utils.math.vector.Vector2;
import org.tobi29.scapes.engine.utils.math.vector.Vector2d;

public class GuiComponentEvent {
    private final double x, y, relativeX, relativeY;
    private final boolean screen;
    private final Vector2 size;

    public GuiComponentEvent() {
        this(Double.NaN, Double.NaN);
    }

    public GuiComponentEvent(double x, double y) {
        this(x, y, Double.NaN, Double.NaN);
    }

    public GuiComponentEvent(double x, double y, double relativeX,
            double relativeY) {
        this(x, y, relativeX, relativeY, true);
    }

    public GuiComponentEvent(double x, double y, double relativeX,
            double relativeY, boolean screen) {
        this(x, y, relativeX, relativeY, screen,
                new Vector2d(Integer.MAX_VALUE, Integer.MAX_VALUE));
    }

    public GuiComponentEvent(double x, double y, Vector2 size) {
        this(x, y, Double.NaN, Double.NaN, size);
    }

    public GuiComponentEvent(double x, double y, double relativeX,
            double relativeY, Vector2 size) {
        this(x, y, relativeX, relativeY, true, size);
    }

    public GuiComponentEvent(double x, double y, double relativeX,
            double relativeY, boolean screen, Vector2 size) {
        this.x = x;
        this.y = y;
        this.relativeX = relativeX;
        this.relativeY = relativeY;
        this.screen = screen;
        this.size = size;
    }

    public GuiComponentEvent(GuiComponentEvent parent, Vector2 size) {
        this(parent.x, parent.y, parent.relativeX, parent.relativeY,
                parent.screen, size);
    }

    public GuiComponentEvent(GuiComponentEvent parent, double x, double y,
            Vector2 size) {
        this(x, y, parent.relativeX, parent.relativeY, parent.screen, size);
    }

    public double x() {
        return x;
    }

    public double y() {
        return y;
    }

    public double relativeX() {
        return relativeX;
    }

    public double relativeY() {
        return relativeY;
    }

    public boolean screen() {
        return screen;
    }

    public Vector2 size() {
        return size;
    }
}

File: Engine/src/main/kotlin/org/tobi29/scapes/engine/gui/GuiComponentEvent.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("UNCHECKED_CAST")

package org.tobi29.scapes.engine.gui

import org.tobi29.scapes.engine.input.ScrollDelta
import org.tobi29.scapes.engine.math.vector.Vector2d

open class GuiComponentEvent(val x: Double = Double.NaN,
                             val y: Double = Double.NaN,
                             val scaleX: Double = 1.0,
                             val scaleY: Double = 1.0,
                             val size: Vector2d = Vector2d(
                                     Int.MAX_VALUE.toDouble(),
                                     Int.MAX_VALUE.toDouble())) {
    open fun copyImpl(
            x: Double,
            y: Double,
            scaleX: Double,
            scaleY: Double,
            size: Vector2d
    ) = GuiComponentEvent(x, y, scaleX, scaleY, size)
}

class GuiComponentEventDrag(
        x: Double = Double.NaN,
        y: Double = Double.NaN,
        scaleX: Double = 1.0,
        scaleY: Double = 1.0,
        val relativeX: Double = Double.NaN,
        val relativeY: Double = Double.NaN,
        size: Vector2d = Vector2d(
                Int.MAX_VALUE.toDouble(),
                Int.MAX_VALUE.toDouble())
) : GuiComponentEvent(x, y, scaleX, scaleY, size) {
    override fun copyImpl(
            x: Double,
            y: Double,
            scaleX: Double,
            scaleY: Double,
            size: Vector2d
    ) = GuiComponentEventDrag(x, y, scaleX, scaleY, relativeX, relativeY, size)
}

class GuiComponentEventScroll(
        x: Double = Double.NaN,
        y: Double = Double.NaN,
        scaleX: Double = 1.0,
        scaleY: Double = 1.0,
        val delta: ScrollDelta,
        size: Vector2d = Vector2d(
                Int.MAX_VALUE.toDouble(),
                Int.MAX_VALUE.toDouble())
) : GuiComponentEvent(x, y, scaleX, scaleY, size) {
    override fun copyImpl(
            x: Double,
            y: Double,
            scaleX: Double,
            scaleY: Double,
            size: Vector2d
    ) = GuiComponentEventScroll(x, y, scaleX, scaleY, delta, size)
}

@Suppress("UNCHECKED_CAST")
fun <T : GuiComponentEvent> T.copy(
        x: Double = this.x,
        y: Double = this.y,
        scaleX: Double = this.scaleX,
        scaleY: Double = this.scaleY,
        size: Vector2d = this.size
) = copyImpl(x, y, scaleX, scaleY, size) as T

@Suppress("UNCHECKED_CAST")
fun <T : GuiComponentEvent> T.scale(
        scale: Vector2d,
        size: Vector2d = this.size
) = copy(x * scale.x, y * scale.y, scaleX * scale.x, scaleY * scale.y, size)
